iff --git a/src/game/behaviors/bowser.inc.c b/src/game/behaviors/bowser.inc.c
index 85bfeb6..127708b 100644
--- a/src/game/behaviors/bowser.inc.c
+++ b/src/game/behaviors/bowser.inc.c
@@ -59,17 +59,19 @@ void (*sBowserTailAnchorActions[])(void) = {
  * Bowser's tail main loop
  */
 void bhv_bowser_tail_anchor_loop(void) {
-    // Call its actions
-    cur_obj_call_action_function(sBowserTailAnchorActions);
-    // Position the tail
-    o->oParentRelativePosX = 90.0f;
+    if (gCurrLevelNum == LEVEL_BOWSER_3) { // STAR ROAD added this check
+        // Call its actions
+        cur_obj_call_action_function(sBowserTailAnchorActions);
+        // Position the tail
+        o->oParentRelativePosX = 90.0f;
+
+        // Make it intangible while Bowser is dead
+        if (o->parentObj->oAction == BOWSER_ACT_DEAD) {
+            o->parentObj->oIntangibleTimer = -1;
+        }

-    // Make it intangible while Bowser is dead
-    if (o->parentObj->oAction == BOWSER_ACT_DEAD) {
-        o->parentObj->oIntangibleTimer = -1;
+        o->oInteractStatus = 0;
     }
-
-    o->oInteractStatus = 0;
 }
@@ -230,15 +232,17 @@ s32 bowser_set_anim_look_down_stop_walk(void) {
  * CamAct changes value on the cutscene itself (cutscene_bowser_arena)
  */
 void bowser_init_camera_actions(void) {
-    if (o->oBowserCamAct == BOWSER_CAM_ACT_IDLE) {
-        o->oAction = BOWSER_ACT_WAIT;
-    } else if (o->oBowserCamAct == BOWSER_CAM_ACT_WALK) {
-        o->oAction = BOWSER_ACT_INTRO_WALK;
-    // Start with a big jump in BitFS to do a platform tilt
-    } else if (o->oBehParams2ndByte == BOWSER_BP_BITFS) {
-        o->oAction = BOWSER_ACT_BIG_JUMP;
-    } else {
-        o->oAction = BOWSER_ACT_DEFAULT;
+    if (gCurrLevelNum == LEVEL_BOWSER_3) { // custom check
+        if (o->oBowserCamAct == BOWSER_CAM_ACT_IDLE) {
+            o->oAction = BOWSER_ACT_WAIT;
+        } else if (o->oBowserCamAct == BOWSER_CAM_ACT_WALK) {
+            o->oAction = BOWSER_ACT_INTRO_WALK;
+        // Start with a big jump in BitFS to do a platform tilt
+        } else if (o->oBehParams2ndByte == BOWSER_BP_BITFS) {
+            o->oAction = BOWSER_ACT_BIG_JUMP;
+        } else {
+            o->oAction = BOWSER_ACT_DEFAULT;
+        }
     }
 }

@@ -1512,6 +1516,10 @@ s32 bowser_check_fallen_off_stage(void) {
     return FALSE;
 }

+#ifdef PLATFORM_DISPLACEMENT_2
+struct PlatformDisplacementInfo sBowserDisplacementInfo;
+#endif
+
 /**
  * Set Bowser's actions
  */
@@ -1590,9 +1598,16 @@ void bowser_free_update(void) {
     UNUSED f32 floorHeight;

     // Platform displacement check (for BitFS)
+#ifdef PLATFORM_DISPLACEMENT_2
+    if ((platform = o->platform) != NULL) {
+        apply_platform_displacement(&sBowserDisplacementInfo, &o->oPosX, (s16 *) &o->oFaceAngleYaw, platform);
+    }
+#else
     if ((platform = o->platform) != NULL) {
         apply_platform_displacement(FALSE, platform);
     }
+#endif
+
     // Reset grabbed status
     o->oBowserGrabbedStatus = BOWSER_GRAB_STATUS_NONE;
     // Update positions and actions (default action)
@@ -1780,6 +1795,10 @@ void bhv_bowser_init(void) {
     // Set health and rainbow light depending of the level
     o->oBowserRainbowLight = sBowserRainbowLight[level];
     o->oHealth = sBowserHealth[level];
+    // custom check (hack so n64 doesn't crash)
+    if (gCurrLevelNum != LEVEL_BOWSER_3) {
+        o->oAnimations = NULL;
+    }
     // Start camera event, this event is not defined so maybe
     // the "start arena" cutscene was originally called this way
     cur_obj_start_cam_event(o, CAM_EVENT_BOWSER_INIT);
diff --git a/src/game/behaviors/breakable_box.inc.c b/src/game/behaviors/breakable_box.inc.c





diff --git a/src/game/behaviors/bully.inc.c b/src/game/behaviors/bully.inc.c
index 6d5fc5c..d3a807b 100644
--- a/src/game/behaviors/bully.inc.c
+++ b/src/game/behaviors/bully.inc.c
@@ -214,9 +214,19 @@ void bully_act_level_death(void) {
             spawn_mist_particles();

             if (o->oBullySubtype == BULLY_STYPE_CHILL) {
+                #ifdef RM2C
+                spawn_default_star(ChillBullyStarPos);
+                #else
                 spawn_default_star(130.0f, 1600.0f, -4335.0f);
+                #endif
             } else {
+                #ifdef RM2C
+                // spawn_default_star(BigBullyStarPos);
+                // STAR ROAD uses this big bully as a Bowser Boss
+                spawn_object(gMarioObject, MODEL_BOWSER_KEY, bhvBowserKey);
+                #else
                 spawn_default_star(0, 950.0f, -6800.0f);
+                #endif
                 spawn_object_abs_with_rot(o, 0, MODEL_NONE, bhvLllTumblingBridge,
                                           0, 154, -5631, 0, 0, 0);
             }






diff --git a/src/game/behaviors/castle_cannon_grate.inc.c b/src/game/behaviors/castle_cannon_grate.inc.c
index 64bd924..5e84620 100644
--- a/src/game/behaviors/castle_cannon_grate.inc.c
+++ b/src/game/behaviors/castle_cannon_grate.inc.c
@@ -1,7 +1,7 @@
 // castle_cannon_grate.inc.c

 void bhv_castle_cannon_grate_init(void) {
-    if (save_file_get_total_star_count(gCurrSaveFileNum - 1, COURSE_MIN - 1, COURSE_MAX - 1) >= 120) {
+    if (save_file_get_total_star_count(gCurrSaveFileNum - 1, COURSE_MIN - 1, COURSE_MAX - 1) >= 65) { // STAR ROAD change
         o->activeFlags = ACTIVE_FLAG_DEACTIVATED;
     }
 }









diff --git a/src/game/behaviors/door.inc.c b/src/game/behaviors/door.inc.c
index c25626c..19ce873 100644
--- a/src/game/behaviors/door.inc.c
+++ b/src/game/behaviors/door.inc.c
@@ -92,21 +92,21 @@ void bhv_door_init(void) {

     x = o->oPosX;
     z = o->oPosZ;
-    find_floor(x, o->oPosY, z, &floor);
+    find_room_floor(x, o->oPosY, z, &floor);
     if (floor != NULL) {
         o->oDoorUnkF8 = floor->room;
     }

     x = o->oPosX + sins(o->oMoveAngleYaw) * 200.0f;
     z = o->oPosZ + coss(o->oMoveAngleYaw) * 200.0f;
-    find_floor(x, o->oPosY, z, &floor);
+    find_room_floor(x, o->oPosY, z, &floor);
     if (floor != NULL) {
         o->oDoorUnkFC = floor->room;
     }

     x = o->oPosX + sins(o->oMoveAngleYaw) * -200.0f;
     z = o->oPosZ + coss(o->oMoveAngleYaw) * -200.0f;
-    find_floor(x, o->oPosY, z, &floor);
+    find_room_floor(x, o->oPosY, z, &floor);
     if (floor != NULL) {
         o->oDoorUnk100 = floor->room;
     }





diff --git a/src/game/behaviors/dorrie.inc.c b/src/game/behaviors/dorrie.inc.c
index 8919677..1316e5b 100644
--- a/src/game/behaviors/dorrie.inc.c
+++ b/src/game/behaviors/dorrie.inc.c
@@ -24,7 +24,7 @@ void dorrie_act_move(void) {

     if (o->oDorrieForwardDistToMario < 320.0f && o->oDorrieGroundPounded) {
         cur_obj_play_sound_2(SOUND_OBJ_DORRIE);
-        o->collisionData = segmented_to_virtual(dorrie_seg6_collision_0600FBB8);
+        //o->collisionData = segmented_to_virtual(dorrie_seg6_collision_0600FBB8);
         o->oAction = DORRIE_ACT_LOWER_HEAD;
         o->oForwardVel = 0.0f;
         o->oDorrieYawVel = 0;
@@ -91,7 +91,7 @@ void dorrie_act_lower_head(void) {
 }

 void dorrie_act_raise_head(void) {
-    o->collisionData = segmented_to_virtual(dorrie_seg6_collision_0600F644);
+    //o->collisionData = segmented_to_virtual(dorrie_seg6_collision_0600F644);
     if (cur_obj_check_if_near_animation_end()) {
         o->oAction = DORRIE_ACT_MOVE;
     } else if (o->oDorrieLiftingMario && o->header.gfx.animInfo.animFrame < 74) {
@@ -108,6 +108,13 @@ void dorrie_act_raise_head(void) {
 }

 void bhv_dorrie_update(void) {
+    // STAR ROAD custom code, copy of bhv_castle_cannon_grate_init
+    // Used along with the 30 Star Door behavior
+    if (save_file_get_total_star_count(gCurrSaveFileNum - 1, COURSE_MIN - 1, COURSE_MAX - 1) >= 30) {
+        o->activeFlags = ACTIVE_FLAG_DEACTIVATED;
+    }
+
+    /*
     f32 boundsShift;
     UNUSED u8 filler[8];
     f32 maxOffsetY;
@@ -167,4 +174,5 @@ void bhv_dorrie_update(void) {

         obj_perform_position_op(1);
     }
+    */
 }











diff --git a/src/game/behaviors/fire_piranha_plant.inc.c b/src/game/behaviors/fire_piranha_plant.inc.c
index be1ea9c..d4e448b 100644
--- a/src/game/behaviors/fire_piranha_plant.inc.c
+++ b/src/game/behaviors/fire_piranha_plant.inc.c
@@ -69,7 +69,13 @@ static void fire_piranha_plant_act_hide(void) {

             if ((u16)(o->oBehParams >> 16) != 0 && o->oHealth == 0) {
                 if (++sNumKilledFirePiranhaPlants == 5) {
-                    spawn_default_star(-6300.0f, -1850.0f, -6300.0f);
+                                       //#ifdef RM2C
+                    //spawn_default_star(BigPiranhasStarPos);
+                                       //#else
+                    //spawn_default_star(-6300.0f, -1850.0f, -6300.0f);
+                                       //#endif
+                    // STAR ROAD uses these piranhas as a Bowser Boss
+                    spawn_object(o, MODEL_BOWSER_KEY, bhvBowserKey);
                 }

                 obj_die_if_health_non_positive();






diff --git a/src/game/behaviors/hidden_star.inc.c b/src/game/behaviors/hidden_star.inc.c
index 9dd2c1e..033bb5c 100644
--- a/src/game/behaviors/hidden_star.inc.c
+++ b/src/game/behaviors/hidden_star.inc.c
@@ -59,7 +59,7 @@ void bhv_bowser_course_red_coin_star_loop(void) {

     switch (o->oAction) {
         case 0:
-            if (o->oHiddenStarTriggerCounter == 8) {
+            if (o->oHiddenStarTriggerCounter == REDS_REQ) {
                 o->oAction = 1;
             }
             break;





diff --git a/src/game/behaviors/king_bobomb.inc.c b/src/game/behaviors/king_bobomb.inc.c
index f867cdf..e2f7133 100644
--- a/src/game/behaviors/king_bobomb.inc.c
+++ b/src/game/behaviors/king_bobomb.inc.c
@@ -30,7 +30,7 @@ void king_bobomb_act_0(void) {
         gSecondCameraFocus = o;
         cur_obj_init_animation_with_sound(5);
         cur_obj_set_pos_to_home();
-        o->oHealth = 3;
+        o->oHealth = KING_BOMB_HEALTH;

         if (cur_obj_can_mario_activate_textbox_2(500.0f, 100.0f)) {
             o->oSubAction++;
@@ -76,8 +76,8 @@ void king_bobomb_act_2(void) {
         }

         if (o->oKingBobombUnk108 == 0) {
-            o->oForwardVel = 3.0f;
-            cur_obj_rotate_yaw_toward(o->oAngleToMario, 0x100);
+            o->oForwardVel = KING_BOMB_FVEL;
+            cur_obj_rotate_yaw_toward(o->oAngleToMario, KING_BOMB_YAWVEL);
         } else {
             o->oForwardVel = 0.0f;
             o->oKingBobombUnk108--;
@@ -214,11 +214,10 @@ void king_bobomb_act_7(void) {
         spawn_mist_particles_variable(0, 0, 200.0f);
         spawn_triangle_break_particles(20, MODEL_DIRT_ANIMATION, 3.0f, 4);
         cur_obj_shake_screen(SHAKE_POS_SMALL);
-#ifndef VERSION_JP
-        cur_obj_spawn_star_at_y_offset(2000.0f, 4500.0f, -4500.0f, 200.0f);
+#ifdef RM2C
+        cur_obj_spawn_star_at_y_offset(KingBobOmbStarPos, 200.0f);
 #else
-        o->oPosY += 100.0f;
-        spawn_default_star(2000.0f, 4500.0f, -4500.0f);
+        cur_obj_spawn_star_at_y_offset(2000.0f, 4500.0f, -4500.0f, 200.0f);
 #endif
         o->oAction = 8;
     }








diff --git a/src/game/behaviors/klepto.inc.c b/src/game/behaviors/klepto.inc.c
index 3cf3369..139b4c2 100644
--- a/src/game/behaviors/klepto.inc.c
+++ b/src/game/behaviors/klepto.inc.c
@@ -329,6 +329,12 @@ void obj_set_speed_to_zero(void) {
 }

 void bhv_klepto_update(void) {
+    // STAR ROAD custom code, copy of bhv_castle_cannon_grate_init
+    if (save_file_get_total_star_count(gCurrSaveFileNum - 1, COURSE_MIN - 1, COURSE_MAX - 1) >= 1) { // STAR ROAD change
+        o->activeFlags = ACTIVE_FLAG_DEACTIVATED;
+    }
+
+    /*
     UNUSED u8 filler[4];

     cur_obj_update_floor_and_walls();
@@ -377,7 +383,11 @@ void bhv_klepto_update(void) {
             || o->oAnimState == KLEPTO_ANIM_STATE_HOLDING_TRANSPARENT_STAR
             #endif
             ) {
+                               #ifdef RM2C
+                spawn_default_star(KleptoStarPos);
+                               #else
                 spawn_default_star(-5550.0f, 300.0f, -930.0f);
+                #endif
             }

             o->oAnimState = KLEPTO_ANIM_STATE_HOLDING_NOTHING;
@@ -399,4 +409,5 @@ void bhv_klepto_update(void) {

     obj_roll_to_match_yaw_turn(o->oKleptoYawToTarget, 0x3000, 600);
     cur_obj_move_standard(78);
+    */
 }








@@ -618,11 +627,11 @@ static void koopa_the_quick_act_race(void) {
                         && (o->oPathedPrevWaypointFlags & WAYPOINT_MASK_00FF) < 28) {
                         // Move faster if mario has already finished the race or
                         // cheated by shooting from cannon
-                        o->oKoopaAgility = 8.0f;
+                        o->oKoopaAgility = KOOPA_SPEED_RACE_END;
                     } else if (o->oKoopaTheQuickRaceIndex != KOOPA_THE_QUICK_BOB_INDEX) {
-                        o->oKoopaAgility = 6.0f;
+                        o->oKoopaAgility = KOOPA_SPEED_THI;
                     } else {
-                        o->oKoopaAgility = 4.0f;
+                        o->oKoopaAgility = KOOPA_SPEED_BOB;
                     }

                     obj_forward_vel_approach(o->oKoopaAgility * 6.0f * downhillSteepness,










diff --git a/src/game/behaviors/mad_piano.inc.c b/src/game/behaviors/mad_piano.inc.c
index 70d816d..e1f29f0 100644
--- a/src/game/behaviors/mad_piano.inc.c
+++ b/src/game/behaviors/mad_piano.inc.c
@@ -59,7 +59,15 @@ static void mad_piano_act_attack(void) {
     cur_obj_move_standard(78);
 }

+#include "game/sound_init.h"
+
 void bhv_mad_piano_update(void) {
+    // STAR ROAD custom code, plays the SMG2 Starship Mario theme
+    if (save_file_get_total_star_count(gCurrSaveFileNum - 1, COURSE_MIN - 1, COURSE_MAX - 1) >= 130) {
+        set_background_music(0, 0x2F, 0);
+    }
+
+    /*
     if (!(o->activeFlags & ACTIVE_FLAG_IN_DIFFERENT_ROOM)) {
         o->oFaceAngleYaw = o->oMoveAngleYaw - 0x4000;

@@ -72,4 +80,5 @@ void bhv_mad_piano_update(void) {
                 break;
         }
     }
+    */
 }






diff --git a/src/game/behaviors/mips.inc.c b/src/game/behaviors/mips.inc.c
index 0aed1ec..17216da 100644
--- a/src/game/behaviors/mips.inc.c
+++ b/src/game/behaviors/mips.inc.c
@@ -12,7 +12,7 @@ void bhv_mips_init(void) {
     u8 starFlags = save_file_get_star_flags(gCurrSaveFileNum - 1, COURSE_NUM_TO_INDEX(COURSE_NONE));

     // If the player has >= 15 stars and hasn't collected first MIPS star...
-    if (save_file_get_total_star_count(gCurrSaveFileNum - 1, COURSE_MIN - 1, COURSE_MAX - 1) >= 15
+    if (save_file_get_total_star_count(gCurrSaveFileNum - 1, COURSE_MIN - 1, COURSE_MAX - 1) >= MIPS1_STAR_REQ
         && !(starFlags & SAVE_FLAG_TO_STAR_FLAG(SAVE_FLAG_COLLECTED_MIPS_STAR_1))) {
         o->oBehParams2ndByte = 0;
 #ifndef VERSION_JP
@@ -20,7 +20,7 @@ void bhv_mips_init(void) {
 #endif
     }
     // If the player has >= 50 stars and hasn't collected second MIPS star...
-    else if (save_file_get_total_star_count(gCurrSaveFileNum - 1, COURSE_MIN - 1, COURSE_MAX - 1) >= 50
+    else if (save_file_get_total_star_count(gCurrSaveFileNum - 1, COURSE_MIN - 1, COURSE_MAX - 1) >= MIPS2_STAR_REQ
              && !(starFlags & SAVE_FLAG_TO_STAR_FLAG(SAVE_FLAG_COLLECTED_MIPS_STAR_2))) {
         o->oBehParams2ndByte = 1;
 #ifndef VERSION_JP




diff --git a/src/game/behaviors/sliding_platform_2.inc.c b/src/game/behaviors/sliding_platform_2.inc.c
index 853bc40..7ae5ea7 100644
--- a/src/game/behaviors/sliding_platform_2.inc.c
+++ b/src/game/behaviors/sliding_platform_2.inc.c
@@ -5,7 +5,7 @@ static Collision const *sSlidingPlatform2CollisionData[] = {
     bits_seg7_collision_0701AA0C,
     bitfs_seg7_collision_07015714,
     bitfs_seg7_collision_07015768,
-    rr_seg7_collision_070295F8,
+    NULL,
     rr_seg7_collision_0702967C,
     NULL,
     bitdw_seg7_collision_0700F688,







diff --git a/src/game/behaviors/whomp.inc.c b/src/game/behaviors/whomp.inc.c
index 3b65ef2..e544dbd 100644
--- a/src/game/behaviors/whomp.inc.c
+++ b/src/game/behaviors/whomp.inc.c
@@ -30,7 +30,7 @@ void whomp_init(void) {
                 seq_player_lower_volume(SEQ_PLAYER_LEVEL, 60, 40);
             } else {
                 cur_obj_set_pos_to_home();
-                o->oHealth = 3;
+                o->oHealth = KING_WHOMP_HEALTH;
             }
         } else if (cur_obj_update_dialog_with_cutscene(MARIO_DIALOG_LOOK_UP,
             DIALOG_FLAG_TURN_TO_MARIO, CUTSCENE_DIALOG, DIALOG_114)) {
@@ -252,7 +252,11 @@ void whomp_die(void) {
             spawn_triangle_break_particles(20, MODEL_DIRT_ANIMATION, 3.0f, 4);
             cur_obj_shake_screen(SHAKE_POS_SMALL);
             o->oPosY += 100.0f;
+                       #ifdef RM2C
+            spawn_default_star(KingWhompStarPos);
+                       #else
             spawn_default_star(180.0f, 3880.0f, 340.0f);
+                       #endif
             cur_obj_play_sound_2(SOUND_OBJ_KING_WHOMP_DEATH);
             o->oAction = 9;


